File: src/signposts.c
#include "signposts.h"
#include <os/signpost.h>

static os_log_t points_of_interest_logger = NULL;

static os_log_t GetPointsOfInterestLogger() {
  if (!points_of_interest_logger) {
    points_of_interest_logger =
        os_log_create("dev.flutter", OS_LOG_CATEGORY_POINTS_OF_INTEREST);
  }
  return points_of_interest_logger;
}

int32_t flt_signposts_emit(const char* str) {
  if (__builtin_available(macOS 10.14, iOS 12.0, *)) {
    os_signpost_event_emit(GetPointsOfInterestLogger(), OS_SIGNPOST_ID_EXCLUSIVE, "flutter", "%s", str);
    return 0;
  }
  return 1;
}


int32_t flt_signposts_begin_interval(uint64_t identifier, const char* str) {
  if (__builtin_available(macOS 10.14, iOS 12.0, *)) {
    os_signpost_interval_begin(GetPointsOfInterestLogger(), identifier, "flutter", "%s", str);
    return 0;
  }
  return 1;
}

int32_t flt_signposts_end_interval(uint64_t identifier, const char* str) {
  if (__builtin_available(macOS 10.14, iOS 12.0, *)) {
    os_signpost_interval_end(GetPointsOfInterestLogger(), identifier, "flutter", "%s", str);
    return 0;
  }
  return 1;
}

File: lib/signposts.dart

import 'dart:async';
import 'dart:ffi';
import 'dart:io';
import 'dart:isolate';
import 'package:ffi/ffi.dart';

import 'signposts_bindings_generated.dart';

/// Emits a Point-of-interest signpost for the "flutter" category.  These will
/// show up in Instruments with the "Points of interest" instrument.
void emitEvent(String msg) {
  int result = _bindings.flt_signposts_emit(msg.toNativeUtf8().cast());
  if (result != 0) {
    throw Exception('emitEvent error: $result');
  }
}

/// Starts recording a region.
/// see also: [endInterval]
void beginInterval(int identifier, String msg) {
  int result = _bindings.flt_signposts_begin_interval(identifier, msg.toNativeUtf8().cast());
  if (result != 0) {
    throw Exception('beginInterval error: $result');
  }
}

/// Ends recording a region (previously started with beginInterval).
/// see also: [beginInterval]
void endInterval(int identifier, String msg) {
  int result = _bindings.flt_signposts_end_interval(identifier, msg.toNativeUtf8().cast());
  if (result != 0) {
    throw Exception('beginInterval error: $result');
  }
}

/// Convenience wrapper for [beginInterval] and [endInterval].
class Interval {
  static int _count = 1;
  Interval(String msg) : identifier = _count++ {
    beginInterval(identifier, msg);
  }
  final int identifier;
  void end(String msg) {
    endInterval(identifier, msg);
  }
}

const String _libName = 'signposts';

/// The dynamic library in which the symbols for [SignpostsBindings] can be found.
final DynamicLibrary _dylib = () {
  if (Platform.isMacOS || Platform.isIOS) {
    return DynamicLibrary.open('$_libName.framework/$_libName');
  }
  if (Platform.isAndroid || Platform.isLinux) {
    return DynamicLibrary.open('lib$_libName.so');
  }
  if (Platform.isWindows) {
    return DynamicLibrary.open('$_libName.dll');
  }
  throw UnsupportedError('Unknown platform: ${Platform.operatingSystem}');
}();

/// The bindings to the native functions in [_dylib].
final SignpostsBindings _bindings = SignpostsBindings(_dylib);
